v2/internal/coord/brdcst/brdcst_followup.go

```diff
@@ -0,0 +1,185 @@
+package brdcst
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/plprobelab/go-kademlia/kad"
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/libp2p/go-libp2p-kad-dht/v2/internal/coord/coordt"
+	"github.com/libp2p/go-libp2p-kad-dht/v2/internal/coord/query"
+	"github.com/libp2p/go-libp2p-kad-dht/v2/tele"
+)
+
+type FollowUp[K kad.Key[K], N kad.NodeID[K], M coordt.Message] struct {
+	queryID query.QueryID
+	pool    *query.Pool[K, N, M]
+	msg     M
+	closest []N
+
+	// nodes we still need to store records with
+	todo    map[string]N
+	waiting map[string]N
+	success map[string]N
+	failed  map[string]struct {
+		Node N
+		Err  error
+	}
+}
+
+func NewFollowUp[K kad.Key[K], N kad.NodeID[K], M coordt.Message](qid query.QueryID, pool *query.Pool[K, N, M], msg M) *FollowUp[K, N, M] {
+	return &FollowUp[K, N, M]{
+		queryID: qid,
+		pool:    pool,
+		msg:     msg,
+		todo:    map[string]N{},
+		waiting: map[string]N{},
+		success: map[string]N{},
+		failed: map[string]struct {
+			Node N
+			Err  error
+		}{},
+	}
+}
+
+func (f *FollowUp[K, N, M]) Advance(ctx context.Context, ev BroadcastEvent) (out BroadcastState) {
+	ctx, span := tele.StartSpan(ctx, "FollowUp.Advance", trace.WithAttributes(tele.AttrInEvent(ev)))
+	defer func() {
+		span.SetAttributes(tele.AttrOutEvent(out))
+		span.End()
+	}()
+
+	switch ev := ev.(type) {
+	case *EventBroadcastStart[K, N]:
+		cmd := &query.EventPoolAddFindCloserQuery[K, N]{
+			QueryID:           ev.QueryID,
+			Target:            ev.Target,
+			KnownClosestNodes: ev.KnownClosestNodes,
+		}
+		return f.advancePool(ctx, cmd)
+	case *EventBroadcastStop:
+		// TODO: ...
+	case *EventBroadcastNodeResponse[K, N]:
+		cmd := &query.EventPoolNodeResponse[K, N]{
+			QueryID:     ev.QueryID,
+			NodeID:      ev.NodeID,
+			CloserNodes: ev.CloserNodes,
+		}
+		return f.advancePool(ctx, cmd)
+	case *EventBroadcastNodeFailure[K, N]:
+		cmd := &query.EventPoolNodeFailure[K, N]{
+			QueryID: ev.QueryID,
+			NodeID:  ev.NodeID,
+			Error:   ev.Error,
+		}
+		return f.advancePool(ctx, cmd)
+	case *EventBroadcastStoreRecordSuccess[K, N, M]:
+		delete(f.waiting, ev.NodeID.String())
+		f.success[ev.NodeID.String()] = ev.NodeID
+	case *EventBroadcastStoreRecordFailure[K, N, M]:
+		delete(f.waiting, ev.NodeID.String())
+		f.failed[ev.NodeID.String()] = struct {
+			Node N
+			Err  error
+		}{Node: ev.NodeID, Err: ev.Error}
+	case *EventBroadcastPoll:
+		// ignore, nothing to do
+		return f.advancePool(ctx, &query.EventPoolPoll{})
+	default:
+		panic(fmt.Sprintf("unexpected event: %T", ev))
+	}
+
+	// TODO: remove duplication with below
```

Comment:
You could call advancePool with a Poll event to let the pool advance some work and avoid this duplication

Reply:
refactored logic and now there's no duplication